File: 02_push_swap/push_swap.py
from collections import deque

stack = deque(map(int, input().split()))
stack_a = deque()
stack_b = deque()

# 크기 순으로 index 재설정
for n1 in stack:
	cnt = 0
	for n2 in stack:
		if n1 > n2:
			cnt += 1
	stack_a.append(cnt)

def a_to_b():
	size = len(stack_a)
	scale = 0.000000053 * size * size + 0.03 * size + 14.5
	i = 0
	while len(stack_a):
		top = stack_a[0]
		if top <= i:
			stack_b.appendleft(stack_a.popleft())
			print("pb")
			i += 1
		elif top > i and top <= i + scale:
			stack_b.appendleft(stack_a.popleft())
			print("pb")
			stack_b.rotate(-1)
			print("rb")
			i += 1
		elif top > i + scale:
			stack_a.rotate(-1)
			print("ra")

def b_to_a():
	while len(stack_b):
		big_idx = stack_b.index(max(stack_b))
		mid = len(stack_b) // 2
		if big_idx > mid:
			while big_idx < len(stack_b):
				stack_b.rotate()
				print("rrb")
				big_idx += 1
		elif big_idx <= mid:
			while big_idx > 0:
				stack_b.rotate(-1)
				print("rb")
				big_idx -= 1
		stack_a.appendleft(stack_b.popleft())
		print("pa")

a_to_b()
# print(stack_a)
# print(stack_b)
b_to_a()
# print(stack_a)
# print(stack_b)

File: 02_push_swap/push_swap.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   push_swap.c                                        :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jiyunlee <jiyunlee@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/02/02 06:03:48 by jiyunlee          #+#    #+#             */
/*   Updated: 2023/02/06 15:18:28 by jiyunlee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "push_swap.h"

void	a_to_b(t_stack *a, t_stack *b);
void	b_to_a(t_stack *a, t_stack *b);

void check_leak() {
	system("leaks -quiet push_swap");
}

int	main(int argc, char *argv[])
{
	t_stack	a;
	t_stack	b;
	char	**arr;

	//atexit(check_leak);
	if (argc == 1)
		return (0);
	else if (argc == 2)
	{
		arr = split_space(argv[1]);
		stack_init(&a, &b, arr);
		free_arr(arr);
	}
	else
		stack_init(&a, &b, ++argv);
	if (is_sorted(&a))
		return (0);
	a_to_b(&a, &b);
	b_to_a(&a, &b);
	// if (is_sorted(&a))
	// 	write(1, "OK\n", 3);
	// else
	// 	write(1, "KO\n", 3);
	free_list(&a);
}

void	a_to_b(t_stack *a, t_stack *b)
{
	float	scale;
	int		top;
	int		i;

	scale = 0.000000053 * a->len * a->len + 0.03 * a->len + 14.5;
	i = 0;
	while (a->len)
	{
		top = a->top->idx;
		if (top <= i)
		{
			push(a, b);
			i++;
		}
		else if (i < top && top <= i + scale)
		{
			push(a, b);
			rotate(b);
			i++;
		}
		else if (top > i + scale)
		{
			if (i < a->len / 2 && i >= 0)
				rev_rotate(a);
			else
				rotate(a);
		}
	}
}

int	get_big_idx(t_stack *stack)
{
	t_list	*tmp;
	int		i;

	tmp = stack->top;
	i = 0;
	while (1)
	{
		if (tmp->idx == stack->len - 1)
			break ;
		tmp = tmp->next;
		i++;
	}
	return (i);
}

void	b_to_a(t_stack *a, t_stack *b)
{
	int	big_idx;
	int	mid;

	while (b->len)
	{
		big_idx = get_big_idx(b);
		mid = b->len / 2;
		if (big_idx > mid)
		{
			while (big_idx < b->len)
			{
				rev_rotate(b);
				big_idx++;
			}
		}
		else if (big_idx <= mid)
		{
			while (big_idx > 0)
			{
				rotate(b);
				big_idx--;
			}
		}
		push(b, a);
	}
}
